src/main/kotlin/com/terraformation/backend/report/db/ReportStore.kt

```diff
@@ -220,6 +221,25 @@ class ReportStore(
     }
   }
 
+  fun delete(reportId: ReportId) {
+    requirePermissions { deleteReport(reportId) }
+
+    // Inform the system that we're about to delete the report and that any external resources tied
+    // to it should be cleaned up.
+    //
+    // This is not wrapped in a transaction because listeners are expected to delete external
+    // resources and then update the database to remove the references to them; if that happened
+    // inside an enclosing transaction, then a listener throwing an exception could cause the system
+    // to roll back the updates that recorded the successful removal of external resources by an
+    // earlier one.
+    //
+    // There's an unavoidable tradeoff here: if a listener fails, the report data will end up
+    // partially deleted.
+    eventPublisher.publishEvent(ReportDeletionStartedEvent(reportId))
+
+    reportsDao.deleteById(reportId)
```

Comment:
Is there a possible race condition here, in which the event listener attempts to `fetchPhotosByReportId` and `fetchFilesByReportId` but the report has already been deleted from the database at this line?

Reply:
Event handlers are called synchronously, so that won't be a problem. `eventPublisher.publishEvent()` won't return until all the handlers have finished.

Reply:
👍 Thanks!